### ISSUE TITLE
Problems with PVTs

### ISSUE THREAD

**Issue by dlmorenob** [OPM/opm-simulators#568]
Dear all,

I have an inconvenient working on flow.  I have managed to configure almost all of a case on the simulator.  I have extended the model to three phases (it was an oil/water problem) including Krs and so on.  However,  my PVT is generating some convergence issues and never completes a time step,  in order to run it I had to use the PVT from another case and then  it works.  The original datafile runs smoothly in Eclipse but generates the errors I mention on flow.   I am using the keyword PVCDO among others.   Is there something I should be taking care of that is out of the ordinary?  I appreciate any comments you may have on the topic.

Thanks.

**Comment by atgeirr** [comment]
Without knowing the case it is hard to guess what can be the cause. Perhaps the PVT tables you input do not cover the pressure range encountered running the case? Since it works with PVT from a different case, you could try to isolate the error by starting with the working PVT, modifying one small thing at a time, to see at which point it breaks.

**Comment by dlmorenob** [comment]
Hello,

I found an error within my datafile and I thought you might have an idea as to how to fix it.   I am trying to run the Brugge case.  I managed to run it using some other PVT properties but the original ones (I used the ones provided by opm for the spe case 1).  I found out there is a problem with the density for the problem.  Also, I am trying to use the PVCDO keyword with just one line and I am getting some convergence errors.

Could this be related to the extrapolation used for the densities?  does it use any?

This is the error I am obtaining:

Iter   MB(Wat)    MB(Oil)    MB(Gas)     CNV(W)     CNV(O)     CNV(G)   W-FLUX(W)  W-FLUX(O)  W-FLUX(G)
   0  6.186e-06  2.253e-23  0.000e+00  1.626e-01  6.593e-03  0.000e+00  2.724e-03  0.000e+00  0.000e+00
MatrixBlockError [bilu0_decomposition:/home/david/Downloads/opm-febrero-2016/dune-istl/dune/istl/ilu.hh:89]: ILU failed to invert matrix block A[4949][4949]FMatrixError [luDecomposition:/home/david/Downloads/opm-febrero-2016/dune-common/dune/common/densematrix.hh:848]: matrix is singular
Solver convergence failed, restarting solver with new time step (3.267 days).
Substep( 0 ), try with stepsize 3.267 (days).

Do you have any idea as how to fix it? I could provide the datafiles if needed (I attach the three phase model pvt data for the model) 

[THREE_PHASE_MODEL_PROPS.TXT](https://github.com/OPM/opm-autodiff/files/130043/THREE_PHASE_MODEL_PROPS.TXT)

**Comment by dlmorenob** [comment]
I also attach the PVT that manages to run the model.
[PVT.txt](https://github.com/OPM/opm-autodiff/files/130045/PVT.txt)

**Comment by andlaus** [comment]
first of all: there has been quite a big change in the implementation of the PVT relations in flow. (recently == last friday). did you update your repositories since then?

> I am trying to run the Brugge case.

Just wondering: is that case publically available? if yes, I suppose it is a good candidate for inclusion into opm-data, if no, do I need to sign an NDA to get access to it?

> Could this be related to the extrapolation used for the densities? does it use any?

that is quite possible: since it uses linear of the first (or last) segment of the table (IIRC), the computed PVT values input values "extremely outside" of the table's range can be complete rubbish (e.g., negative).

**Comment by dlmorenob** [comment]
I did the last update last Friday around 3:00 pm in the afternoon my time.  That would be 10 pm yours.  I think it is the last one.  Here is the result of running it with the correct pvt.

Iter   MB(Wat)    MB(Oil)    MB(Gas)     CNV(W)     CNV(O)     CNV(G)   W-FLUX(W)  W-FLUX(O)  W-FLUX(G)
   0  2.871e-10  1.046e-27  0.000e+00  7.547e-06  3.060e-07  0.000e+00  2.724e-03  0.000e+00  0.000e+00
MatrixBlockError [bilu0_decomposition:/home/david/Downloads/opm-febrero-2016/dune-istl/dune/istl/ilu.hh:89]: ILU failed to invert matrix block A[3972][3972]FMatrixError [luDecomposition:/home/david/Downloads/opm-febrero-2016/dune-common/dune/common/densematrix.hh:848]: matrix is singular
Solver convergence failed, restarting solver with new time step (0.000151626 days).
Substep( 0 ), try with stepsize 0.000151626 (days).

Iter  W-FLUX(W)  W-FLUX(O)  W-FLUX(G)
   0  4.383e-03  6.722e-17  6.381e-17
   1  3.208e-17  0.000e+00  0.000e+00
well converged iter: 1

Iter   MB(Wat)    MB(Oil)    MB(Gas)     CNV(W)     CNV(O)     CNV(G)   W-FLUX(W)  W-FLUX(O)  W-FLUX(G)
   0  9.475e-11  3.450e-28  0.000e+00  2.491e-06  1.010e-07  0.000e+00  2.724e-03  0.000e+00  0.000e+00
MatrixBlockError [bilu0_decomposition:/home/david/Downloads/opm-febrero-2016/dune-istl/dune/istl/ilu.hh:89]: ILU failed to invert matrix block A[3972][3972]FMatrixError [luDecomposition:/home/david/Downloads/opm-febrero-2016/dune-common/dune/common/densematrix.hh:848]: matrix is singular
Program threw an exception: [/home/david/Downloads/opm-febrero-2016/opm-core/opm/core/simulator/AdaptiveTimeStepping_impl.hpp:238] Solver failed to converge after 10 restarts.

With the other one it runs smoothly.  

Yes,  It is the publicly available model.  I have already extended it to three phases and I can provide it to you for inclusion into the datafile even though I find convenient to ask Peters at TNO first.  Please provide me an email and I will send the full dataset.  I could write the how to for the problem for OPM if you want.

David.

**Comment by andlaus** [comment]
> I did the last update last Friday around 3:00 pm in the afternoon my time. That would be 10 pm yours. I think it is the last one. Here is the result of running it with the correct pvt.

that should be fine. did you also update the opm-material module?

> Yes, It is the publicly available model. I have already extended it to three phases and I can provide it to you for inclusion into the datafile even though I find convenient to ask Peters at TNO first. Please provide me an email and I will send the full dataset. I could write the how to for the problem for OPM if you want.

ok, I send you a private email...

**Comment by GitPaean** [comment]
Please try the patch in OPM/opm-parser#692 and tell how it works with your problem.

**Comment by dlmorenob** [comment]
Hello,

The simulation is running fine.  Thanks.  May I ask what was the error? was it dividing by zero somewhere?

Thanks.

David

**Comment by GitPaean** [comment]
The reason is the dimension/unit of the compressibility in the keyword PVCDO was wrong, which makes the value of compressiblity really wrong after the unit conversion. PVCDO is not a keyword we have been using very often, so the problem has not been recovered. We had one example with this keyword, while the compressiblity in that case is zero, so it did not matter that much.

A few weeks ago, I got an example with PVCDO and the result was also problematic. From the result I knew it is something related to the PVCDO, while just until yesterday I found time to look into it. 

If your problem was solved completely, feel free to close this issue. 

Best, 
Kai
